### ISSUE TITLE
cfund - accepted proposals which expire will be immediately re-accepted

### ISSUE THREAD

**Issue by marcus290** [NAVCoin/navcoin-core#290]
If an accepted proposal runs past its duration and expires, at the end of the voting cycle, instead of just being marked as expired, it will expire and then be immediately re-accepted.
![screenshot from 2018-09-06 14-28-52](https://user-images.githubusercontent.com/36244342/45131494-3af9af80-b1e1-11e8-8567-b6bd089fb3e2.png)

**Comment by marcus290** [comment]
@aguycalled 
Hey, looks like the order of if-statements here, in main.cpp line 3748:
```
                if(proposal.IsExpired(pindexNew->GetMedianTimePast()) && proposal.fState != CFund::EXPIRED) {
                    if(proposal.fState == CFund::ACCEPTED) {
                        pindexNew->nCFSupply += proposal.GetAvailable();
                        pindexNew->nCFLocked -= proposal.GetAvailable();
                    }
                    proposal.fState = CFund::EXPIRED;
                    cout << "proposal marked as expired!";
                    fUpdate = true;
                }
                else if(proposal.IsRejected() && proposal.fState != CFund::REJECTED) {
                    proposal.fState = CFund::REJECTED;
                    fUpdate = true;
                }

                if(proposal.IsAccepted() && proposal.fState != CFund::ACCEPTED) {
                    if(pindexNew->nCFSupply >= proposal.GetAvailable()) {
                        pindexNew->nCFSupply -= proposal.GetAvailable();
                        pindexNew->nCFLocked += proposal.GetAvailable();
                        proposal.fState = CFund::ACCEPTED;
                        cout << "proposal marked as accepted!";
                        proposal.blockhash = pindexNew->GetBlockHash();
                        fUpdate = true;
                    } else if(proposal.fState != CFund::PENDING_FUNDS) {
                        proposal.fState = CFund::PENDING_FUNDS;
                        fUpdate = true;
                    }
                }
```
what do you think?

**Comment by marcus290** [comment]
Here, the proposal.fState is marked as expired and then immediately marked as accepted.

I tried switching the two if-statements but it does not fix the issue - the expired proposal will still be re-accepted at the end of the next voting cycle.

**Comment by aguycalled** [comment]
I would say this issue its just cosmetic, fState never moves out of ACCEPTED.
The logic error is very likely in Proposal::GetState().

**Comment by aguycalled** [comment]
You can get a better overview of what happens, if you add debug of fState to CProposal::ToJson()

**Comment by marcus290** [comment]
I checked the fState as you suggested and added breakpoints around the if-statements in main.cpp:
```
breakpoint1
                if(proposal.IsExpired(pindexNew->GetMedianTimePast()) && proposal.fState != CFund::EXPIRED) {
                ...
                }
breakpoint2
                if(proposal.IsAccepted() && proposal.fState != CFund::ACCEPTED) {
                ...
                }
breakpoint3
```

At the next end of voting period after a proposal duration expires, the fState will move from ACCEPTED to EXPIRED to ACCEPTED at breakpoints 1, 2, and 3. Also, the expired proposal is re-approved on a new block, which shows the same proposal is being accepted a second time. Do you get the same result?

For an expired proposal, I think proposal.IsAccepted() will incorrectly recount nVotesYes which were from the proposal originally being accepted

**Comment by aguycalled** [comment]
I could now spend some time reviewing this, and fixed the logic making a proposal only able to move to ACCEPTED from a previous state NIL or PENDING_FUNDS. Looks like this is fixed, please reopen if still happens.
afd4bb2
thx @marcus290 for spotting this!

**Comment by marcus290** [comment]
Not sure if there is a problem with my build but I'm finding proposals in regnet which are expired (i.e. fState = 3) will have a status "accepted waiting for end of voting cycle" and be accepted in the next cycle.

```
 listproposals
[
  {
    "version": 2,
    "hash": "0ae3e5e56b4b8c78b372a4181cc18910f0f56a823c9565464c8e4b240143a2b5",
    "blockHash": "29e9c21c5c12014988150a61941197aa9345f2e43777742fc4d2e111a6354237",
    "description": "test-expiry-preq",
    "requestedAmount": "100.00",
    "notPaidYet": "90.00",
    "userPaidFee": "0.0001",
    "paymentAddress": "musvHFSxtxcns89PHHBTvMyKPNVC1EmmHw",
    "proposalDuration": 120,
    "votesYes": 178,
    "votesNo": 0,
    "votingCycle": 0,
    "status": "accepted waiting for end of voting period",
    "state": 3,
    "paymentRequests": [
      {
        "version": 2,
        "hash": "5d178ba34e5335c352a96a90eaa22d2e60a619b92f56c7b12b2e3882e4e1c7ab",
        "blockHash": "7fa54185db35fb93f8ba1ac356cc2f785378d3ebd364ba441144334b7e0bdbbe",
        "description": "test-expired",
        "requestedAmount": "10.00",
        "votesYes": 174,
        "votesNo": 0,
        "votingCycle": 2,
        "status": "accepted",
        "state": 1,
        "stateChangedOnBlock": "b369b97506fd243c2188a6e05c855f2e08c05d2a703b2fe5eb1dc658ecc49e9f",
        "paidOnBlock": "0000000000000000000000000000000000000000000000000000000000000000"
      }
    ]
  }
]
generate 180
listproposals
[
  {
    "version": 2,
    "hash": "0ae3e5e56b4b8c78b372a4181cc18910f0f56a823c9565464c8e4b240143a2b5",
    "blockHash": "29e9c21c5c12014988150a61941197aa9345f2e43777742fc4d2e111a6354237",
    "description": "test-expiry-preq",
    "requestedAmount": "100.00",
    "notPaidYet": "100.00",
    "userPaidFee": "0.0001",
    "paymentAddress": "musvHFSxtxcns89PHHBTvMyKPNVC1EmmHw",
    "proposalDuration": 120,
    "expiresOn": 1544760808,
    "votesYes": 178,
    "votesNo": 0,
    "votingCycle": 0,
    "status": "accepted",
    "state": 1,
    "stateChangedOnBlock": "9a36f0f5e6782233c9c1c49f43a14c2c36cc71e1e9b64157847136268a01997f",
    "paymentRequests": [
      {
        "version": 2,
        "hash": "5d178ba34e5335c352a96a90eaa22d2e60a619b92f56c7b12b2e3882e4e1c7ab",
        "blockHash": "7fa54185db35fb93f8ba1ac356cc2f785378d3ebd364ba441144334b7e0bdbbe",
        "description": "test-expired",
        "requestedAmount": "10.00",
        "votesYes": 174,
        "votesNo": 0,
        "votingCycle": 4,
        "status": "accepted waiting for end of voting period",
        "state": 0,
        "stateChangedOnBlock": "0000000000000000000000000000000000000000000000000000000000000000"
      }
    ]
  }
]
```

**Comment by marcus290** [comment]
Fixed by #367
